components/ide/code/extensionhelper/main.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	yaml "gopkg.in/yaml.v2"
+
+	gitpod "github.com/gitpod-io/gitpod/gitpod-protocol"
+	supervisor "github.com/gitpod-io/gitpod/supervisor/api"
+)
+
+func main() {
+	// Wait until content ready
+	contentStatus, wsInfo, err := resolveWorkspaceInfo(context.Background())
+	if err != nil || wsInfo == nil || contentStatus == nil || !contentStatus.Available {
+		return
+	}
+	output := "--start-server"
+	if strings.Contains(wsInfo.GetWorkspaceContextUrl(), "github.com") {
+		output += " --install-builtin-extension github.vscode-pull-request-github"
+	}
+	uniqMap := map[string]struct{}{}
+	extensions, _ := getExtensions(wsInfo.GetCheckoutLocation())
+	for _, ext := range extensions {
+		if _, ok := uniqMap[ext]; ok {
+			continue
+		}
+		uniqMap[ext] = struct{}{}
+		output += " --install-extension " + ext
+	}
+	fmt.Print(output)
+}
+
+func getExtensions(repoRoot string) (extensions []string, err error) {
+	if repoRoot == "" {
+		return
+	}
+	data, err := os.ReadFile(filepath.Join(repoRoot, ".gitpod.yml"))
+	if err != nil {
+		return
+	}
+	var config *gitpod.GitpodConfig
+	if err = yaml.Unmarshal(data, &config); err != nil {
+		return
+	}
+	if config == nil || config.Vscode == nil {
+		return
+	}
+	var wg sync.WaitGroup
+	var extensionsMu sync.Mutex
+	for _, ext := range config.Vscode.Extensions {
+		lowerCaseExtension := strings.ToLower(ext)
+		if isUrl(lowerCaseExtension) {
+			wg.Add(1)
+			go func(url string) {
+				defer wg.Done()
+				location, err := downloadExtension(url)
+				if err != nil {
+					return
+				}
+				extensionsMu.Lock()
+				extensions = append(extensions, location)
+				extensionsMu.Unlock()
+			}(ext)
+		} else {
+			path := filepath.Join(repoRoot, ext)
+			if isVsixFileExists(path) {
```

Comment:
@mustard-mh it is something new? I cannot remember such cases in VS Code server customisations

Reply:
> Support install extensions with file path in `.gitpod.yml` file
This, it's easy to add, and in case users want it

Reply:
Let's remove it for now and wait for requirements. I doubt that extensions will belong to workspace folder, but rather in stalled somewhere in /tmp during docker build for workspace.

Reply:
With download func below, files will be downloaded to `/tmp/vsix*.vsix`